### PR TITLE
set env vars at build time through docker build --build_args

### LINKED ISSUE #537: Set build-time variables to avoid tokens in Dockerfile
**Is your feature request related to a problem? Please describe.**
It's not the first time I need to use tokens in the build phase. Right now we need to embed them directly in the Dockerfile. For config variables is not a problem but for tokens or passwords it's preferable to avoid committing them in the Dockerfile.

**Describe the solution you'd like**
Have a section similar to the one for environmental variables but for build args that are passed with `--build-arg` and can be used inside the Dockerfile.

```
ARG SOME_TOKEN
```

**Additional context**
One use case example is setting the `NPM_TOKEN` for private repositories. It's the official solution proposed by NPM to use in Docker environments.

[https://docs.npmjs.com/docker-and-private-modules](https://docs.npmjs.com/docker-and-private-modules)

### PR DESCRIPTION
**Pull request by jpolvora**
fix #537

### CODE REVIEW

File: src/docker/DockerApi.ts [hunk 1]
```diff
@@ -227,7 +227,8 @@ class DockerApi {
         imageName: string,
         newVersionNumber: number,
         tarballFilePath: string,
-        buildLogs: BuildLog
+        buildLogs: BuildLog,
+        buildargs: Object = {}
```
Comment (githubsaturn) [inline src/docker/DockerApi.ts]: Needs to be strongly typed, rather than a loose object.

File: src/user/ImageMaker.ts [hunk 2]
```diff
@@ -74,7 +75,7 @@ export default class ImageMaker {
     constructor(
         private dockerRegistryHelper: DockerRegistryHelper,
         private dockerApi: DockerApi,
-        private namespace: string,
+        private dataStore: DataStore,
```
Comment (githubsaturn) [inline src/user/ImageMaker.ts]: Only pass the env vars as needed - avoid adding  an unnecessary extra dependency on DataStore.

File: src/user/ImageMaker.ts [hunk 3]
```diff
@@ -178,7 +184,8 @@ export default class ImageMaker {
                             tarFilePath,
                             baseImageNameWithoutVerAndReg,
                             appName,
-                            appVersion
+                            appVersion,
+                            appDefinition
```
Comment (githubsaturn) [inline src/user/ImageMaker.ts]: Same here, only need to pass the env vars. Avoid unnecessary injection.

**Comment by jpolvora** [comment]
Thanks for your feedback.
I've made the changes to fit your advices in code.

Sorry for not discussing the the feature.

For a docker novice like me, it was not clear that enviroment vars was available at build time. Today it's clear to me, but it's not so bad have the app env vars available at docker image build time. 

I have an app that is multi -tenant and the build image is different depending on which client I'm dealing with. So, run scripts like CMD ["npm run start-${CLIENT}] it's possible if I define ARG CLIENT and ENV $CLIENT  etc etc.

**Comment by jpolvora** [comment]
adjustments made

**Pull request merged**